File: String to Integer (atoi)/String to Integer (atoi).cpp
class Solution {
public:
    int atoi(const char *str) {
        long long res = 0;
        bool isNegative = false;
        while (*str == ' ') ++str;
        if (*str == '-') {
            isNegative = true;
            ++str;
        } else if (*str == '+') ++str;
        for (; *str >= '0' && *str <= '9'; ++str) {
            res = 10 * res + *str - '0';
        }
        if (isNegative) res = -res;
        if (res > INT_MAX) res = INT_MAX;
        else if (res < INT_MIN) res = INT_MIN;
        return static_cast<int>(res);
    }
};

File: String to Integer (atoi)/String to Integer (atoi).java
public class Solution {
    public int atoi(String str) {
        if (str.length() == 0) return 0;
        char[] c = str.toCharArray();
        long res = 0;
        boolean isNegative = false;
        int i = 0;
        for (; i < c.length && c[i] == ' '; ++i) {
        }
        if (c[i] == '-') {
            isNegative = true;
            ++i;
        } else if (c[i] == '+') ++i;
        for (; i < c.length && c[i] >= '0' && c[i] <= '9'; ++i) {
            res = res * 10 + c[i] - '0';
            if (res > Integer.MAX_VALUE)  break;
        }
        if (isNegative) res = -res;
        if (res > Integer.MAX_VALUE) res = Integer.MAX_VALUE;
        else if (res < Integer.MIN_VALUE) res = Integer.MIN_VALUE;
        return (int)res;
    }
}